File: admin/server/src/models/creditLedger.model.js
const mongoose = require('mongoose');
const { baseSchemaOptions } = require('../utils/baseSchema');

/**
 * Append-only audit log of every try-on credit event. This is the single source
 * of truth for the costing dashboard (how much of the purchased BrandShoot pool
 * has been consumed) and for a user's credit history.
 *
 * `amount` is SIGNED: positive = credits added (grant/refund/purchase),
 * negative = consumed (customer try-on, admin photoshoot/catalog pool draw).
 * `userId` is null for pool-level events (purchases, admin generations) that are
 * not tied to a specific shopper.
 */
const creditLedgerSchema = new mongoose.Schema(
  {
    type: {
      type: String,
      required: true,
      enum: [
        'signup_grant', // free credits when a user account is created
        'admin_grant', // admin added credits to a user
        'admin_set', // admin set a user's balance to an exact value
        'admin_deduct', // admin removed credits from a user
        'tryon_consume', // a customer try-on spent 1 credit
        'tryon_refund', // a failed try-on refunded 1 credit
        'pool_purchase', // admin recorded a purchase of pool credits
        'pool_consume', // admin photoshoot/catalog drew from the pool
      ],
    },
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null, index: true },
    amount: { type: Number, required: true }, // signed
    balanceAfter: { type: Number, default: null }, // user balance after (null for pool events)
    feature: { type: String, default: '' }, // tryon | photoshoot | catalog | signup | admin | purchase
    jobId: { type: String, default: '', index: true }, // BrandShoot job id (traceability + idempotency)
    reason: { type: String, default: '' }, // free-text admin note
    createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null }, // admin who acted
  },
  baseSchemaOptions
);

creditLedgerSchema.index({ userId: 1, createdAt: -1 });
// Prevent double-counting a consume/refund for the same BrandShoot job.
creditLedgerSchema.index(
  { jobId: 1, type: 1 },
  { unique: true, partialFilterExpression: { jobId: { $gt: '' } } }
);

module.exports = mongoose.model('CreditLedger', creditLedgerSchema, 'creditledgers');

File: admin/lib/data/repositories/credits/credit_ledger_model.dart
/// One row of the try-on credit ledger (backend: creditLedger.model.js).
class CreditLedgerModel {
  final String id;
  final String type; // signup_grant | admin_grant | admin_set | admin_deduct | tryon_consume | ...
  final int amount; // signed: + added, - consumed
  final int? balanceAfter;
  final String feature;
  final String reason;
  final DateTime? createdAt;

  CreditLedgerModel({
    this.id = '',
    this.type = '',
    this.amount = 0,
    this.balanceAfter,
    this.feature = '',
    this.reason = '',
    this.createdAt,
  });

  factory CreditLedgerModel.fromJson(Map<String, dynamic> data) {
    DateTime? parseDate(dynamic v) {
      if (v == null) return null;
      return DateTime.tryParse(v.toString());
    }

    return CreditLedgerModel(
      id: (data['id'] ?? data['_id'] ?? '').toString(),
      type: (data['type'] ?? '').toString(),
      amount: (num.tryParse((data['amount'] ?? 0).toString()) ?? 0).toInt(),
      balanceAfter: data['balanceAfter'] == null
          ? null
          : (num.tryParse(data['balanceAfter'].toString()) ?? 0).toInt(),
      feature: (data['feature'] ?? '').toString(),
      reason: (data['reason'] ?? '').toString(),
      createdAt: parseDate(data['CreatedAt'] ?? data['createdAt']),
    );
  }

  /// Human-friendly label for the ledger row's type.
  String get label {
    switch (type) {
      case 'signup_grant':
        return 'Signup bonus';
      case 'admin_grant':
        return 'Admin added';
      case 'admin_set':
        return 'Admin set balance';
      case 'admin_deduct':
        return 'Admin removed';
      case 'tryon_consume':
        return 'Try-on';
      case 'tryon_refund':
        return 'Try-on refund';
      case 'pool_purchase':
        return 'Pool purchase';
      case 'pool_consume':
        return 'Admin generation';
      default:
        return type;
    }
  }
}
